Adam Whitlock Edits v1 (#1)
tylerauerbeck/my-dev.to: kubevirt/K8SVM/K8SVM.md

## The solution!

This is when I ran into KubeVirt. I quickly realized that we had all the access we needed to be able to deploy and run things in the cluster. So after a quick proof-of-concept, I felt fairly confident that we could do what we needed to do all within the cluster. The big challenge at the time was ensuring that we were able to communicate over the network using the protocols that we needed to (primarily SSH required by Ansible). There were likely a ton of better ways to solve this problem and there are absolutely better ways that I know of doing this now. But to avoid having to dig a bit into Kubernetes networking, I'll stick with the simplest <strike>solution</strike> hack that I found to get around the problem at the time. So without further delay, let's get to work.

### Step 1: Find yourself a Kubernetes platform

> While this scenario was originally run on an Openshift cluster, this will work on any Kubernetes platform. To keep things simple, I'll be use [KinD](https://github.com/kubernetes-sigs/kind). The only thing you need to get this running is to grab the KinD release from Github and to have Docker running on your machine. For the purposes of this article, we'll be using KinD version [v0.6.0](https://github.com/kubernetes-sigs/kind/releases/tag/v0.6.0).

Reviewer: ```suggestion
While this scenario was originally run on an Openshift cluster, this will work on any Kubernetes platform. To keep things simple, I'll use [KinD](https://github.com/kubernetes-sigs/kind). The only thing you need to get this running is to grab the KinD release from Github and to have Docker running on your machine. For the purposes of this article, we'll be using KinD version [v0.6.0](https://github.com/kubernetes-sigs/kind/releases/tag/v0.6.0).
```

- role: worker
```

> What this will give us is a 1 master & 1 worker cluster to start working with. All we need to do is deploy this and we'll be ready to get started. You can do this with the following command: `kind create cluster --config cluster.yml --name kv-demo`. After running this command you'll see output similar to the one below:

Reviewer: ```suggestion
What this will give us is a 1 master and 1 worker cluster to start working with. All we need to do now is deploy this and we'll be ready to get started. You can do this with the following command: `kind create cluster --config cluster.yml --name kv-demo`. After running this command you'll see output similar to the one below:
```

### Step 4: Now for some Virtual Machines

> As part of our KubeVirt deployment, we created a handful of CRD's. One of those was called `virtualmachineinstances`. As you may guess, this is one way we have of creating a VM. There are additional methods, but this is the approach we'll take as part of this example. We'll focus on just deploying a single VM for brevity. You can template to your liking (or just modified and run multiple times) in order to create multiple VM's. In order to create our single instance, you can run the following:

Reviewer: ```suggestion
As part of our KubeVirt deployment, we created a handful of CRD's. One of those was called `virtualmachineinstances`. As you may guess, this is one way we have of creating a VM. There are additional methods, but this is the approach we'll take as part of this example. We'll focus on just deploying a single VM for brevity. However, you can customize this into a template to fit your requirements (or even just modify and run multiple times) in order to create multiple VM's. In order to create our single instance, you can run the following:
```

#### Configure your workspace

Now that we have the IP address for where we need to connect, let's make sure that our workspace is in good shape to begin communicating with it. The first thing we need to do is reconnect to our tool-box container. We can do this again by running `kubectl exec -it tool-box-64f5d796-2db66 /bin/bash`. Once we're dropped back in our terminal we need to do a few things
- Retrieve our private key
- Configure an Ansible inventory
- Create a simple Ansible playbook

##### Retrieve Private Key and Test Connectivity

> We need use `curl` to download our private key from the repository noted above (if you chose to use the example-provided public key). We could get fancier and  mount the key in as part of our deployment, but the point here was to keep things as simple as possible to start.

Reviewer: ```suggestion
We need to use `curl` in order to download our private key from the repository noted above (if you chose to use the example-provided public key). We could get fancier and  mount the key in as part of our deployment, but the point here was to keep things as simple as possible to start.
```

      become: true
```

> Once this is ready, we can go ahead and run our playbook, which creates a user named `James` on the VM that we had created. We can confirm this by remote accessing the host via SSH and checking that the user James exists.

Reviewer: ```suggestion
Once this is ready, we can go ahead and run our playbook, which creates a user named `James` on the VM that we had created. We can confirm this by accessing the host via SSH and checking that the user James exists.
```
